README.md
# EasyCAPTCHA

[![Build Status](https://api.travis-ci.com/K-and-R/easy_captcha.svg?branch=master)](https://travis-ci.com/github/K-and-R/easy_captcha)
[![Maintainability](https://api.codeclimate.com/v1/badges/0505012081a472de86e2/maintainability)](https://codeclimate.com/github/K-and-R/easy_captcha/maintainability)
[![Test Coverage](https://api.codeclimate.com/v1/badges/0505012081a472de86e2/test_coverage)](https://codeclimate.com/github/K-and-R/easy_captcha/test_coverage)

A simple captcha implementation for Rails 5+ based on RMagick.

This (`kandr-easy_captcha`) is a fork of
[EasyCaptcha](https://github.com/phatworx/easy_captcha) (`easy_captcha`) with Rails 5+ support. This fork is
maintained by Karl Wilbur/K&R Software (karl@kandrsoftware.com).

Testing is done against current Ruby versions; EOL'd versions are dropped from testing, but should continue to work.

## Dependencies/Requirements

* Rails 5+ (<http://github.com/rails/rails>)
* RMagick

RMagick should be included in your `Gemfile`

```ruby
  gem 'rmagick'
```

for Java/JRuby you can use

```ruby
  gem 'rmagick4j'
```

## Installation

add to Gemfile

```ruby
  gem 'kandr-easy_captcha', require: 'easy_captcha'
```

after running `bundle install`, execute

```bash
  rails g easy_captcha:install
```

## Configuration

You can configure `easy_captcha` in `config/initializers/easy_captcha.rb`, if
you want to customize the default configuration

```ruby
  EasyCaptcha.setup do |config|
    # #####
    # # Cache
    # config.cache                   = true
    # Cache temp dir from Rails.root
    # config.cache_temp_dir          = Rails.root.join('tmp', 'captchas')
    # Cache size
    # config.cache_size              = 500
    # Cache expire
    # config.cache_expire            = 1.day

    # #####
    # # CAPTCHA
    # # Chars available for CAPTCHA
    # config.captcha_character_pool  = %w(2 3 4 5 6 7 9 A C D E F G H J K L M N P Q R S T U X Y Z)
    #
    # # Length of CAPTCHA string
    # config.captcha_character_count = 6
    # config.captcha_character_count_max = 6
    # config.captcha_character_count_min = 6
    #
    # # CAPTCHA Image dimensions
    # config.captcha_image_height    = 40

    #####
    # Image
    # config.captcha_image_height    = 40
    # config.captcha_image_width     = 140

    # #####
    # # eSpeak
    # # Enable eSpeak using all defaults:
    # config.espeak = true

    # # Enable eSpeak using custom config:
    # config.espeak do |espeak|
    #   Amplitude, 0 to 200
    #   espeak.amplitude = 80..120

    #   Word gap. Pause between words
    #   espeak.gap = 80

    #   Pitch adjustment, 0 to 99
    #   espeak.pitch = 30..70

    #   Use voice file of this name from espeak-data/voices
    #   espeak.voice = nil
    # end

    # #####
    # # Generator
    # config.generator :default do |generator|
    #   # # Backgound
    #   # # Ignored (set as `none`) when using custom background image or fill
    #   # generator.background_color = '#FFFFFF'
    #   # An instance of a `Fill` class. RMagick supplies three `Fill` classes:
    #   #    - `HatchFill`
    #   #    - `GradientFill`
    #   #    - `TextureFill`.
    #   generator.background_fill  = nil
    #   # # Custom image to be used for CAPTCHA background
    #   # generator.background_image = nil

    #   # Blur
    #   # generator.blur                   = true
    #   # generator.blur_radius            = 1
    #   # generator.blur_sigma             = 2

    #   # # Font
    #   # generator.font_size              = 24
    #   # generator.font_fill_color        = '#333333'
    #   # generator.font_stroke_color      = '#000000'
    #   # generator.font_stroke            = 0
    #   # generator.font_family            = File.expand_path('../../resources/afont.ttf', __FILE__)

    #   # # Implode  (the "funhouse mirror" effect; negative values allowed)
    #   # generator.implode                = 0.1

    #   # # Sketch
    #   # generator.sketch                 = true
    #   # generator.sketch_radius          = 0.0
    #   # generator.sketch_sigma           = 1.0

    #   # Wave
    #   # generator.wave                   = true
    #   # generator.wave_length            = (60..100)
    #   # generator.wave_amplitude         = (3..5)
    # end
  end
```

## Caching

It is strongly recommended to enable caching. You can see the three paramters which you have to fill in your config
file below.

```ruby
  EasyCaptcha.setup do |config|
    # Cache
    config.cache            = true
    # Cache temp dir from Rails.root
    config.cache_temp_dir   = Rails.root.join('tmp', 'captchas')
    # Cache expire
    config.cache_expire     = 1.day
    # Cache size
    # config.cache_size     = 500
  end
```

## Example

```ruby
  <% form_tag '/' do %>
    <% if request.post? %>
      <p><%= valid_captcha?(params[:captcha]) ? 'valid' : 'invalid' %> captcha</p>
    <% end %>
    <p><%= captcha_tag %></p>
    <p><%= text_field_tag :captcha %></p>
    <p><%= submit_tag 'Validate' %></p>
  <% end %>
```

## Example app

You find an example app under: <http://github.com/phatworx/easy_captcha_example>

## History

See the [CHANGELOG.md](./CHANGELOG.md)

## Maintainers

* Karl Wilbur (<http://github.com/karlwilbur>)
* K&R Software (<http://github.com/K-and-R>)

## Contributing to (K&R) EasyCaptcha

* See the [TODO.md](./TODO.md) file for a list of needed changes
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it!
* Please do not edit `Rakefile`, `EasyCaptcha::VERSION`, or `CHANGELOG`

## Copyright

EasyCAPTCHA is licensed under The MIT License.

Copyright (c) 2010-2021 Marco Scholl; K&R Software, LLC.; and EasyCAPTCHA Contributors.
See [`LICENSE.txt`](./LICENSE.txt) and [`CHANGELOG.md`](./CHANGELOG.md) for further details.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.11.0 (2021-05-04)

* [karlwilbur] - Add support `background_fill` generator config option
* [karlwilbur] - *DEPRECATE* `image_background_color` config option in favor of `background_color`
* [karlwilbur] - Refactor specs:

    - Remove spec options from repo
    - Add specs for default generator

* [karlwilbur] - Remove `Gemfile.lock`
* [karlwilbur] - Updates to `README.md`

## 0.10.0 (2021-02-22)

* [karlwilbur] - Use `kandr-pry-plus` gem instead of the stale and broken `pry-plus`
* [karlwilbur] - Reset CAPTCHA code on subsequent requests for a new image/audio file
* [karlwilbur] - Add motion blur, using same blur values, when blur is used
* [karlwilbur] - Add support for a range of CAPTCHA characters
* [karlwilbur] - Updates to `README.md`

## 0.9.2 (2020-12-16)

* [karlwilbur] - [gemspec] Fix files included in gem
* [karlwilbur] - Add `bin/` script to push coverage report to CodeClimate

## 0.9.1 (2020-12-06)

### Additions

* [karlwilbur] - Add Travis CI badge to `README`

### Modifications/Fixes

* [karlwilbur] - Fix `Magick::Image#destroyed?` method guard syntax for jRuby support
* [karlwilbur] - Loosen Rubygems dependency from `>= 3.0.0` to `>= 2.7.0`
* [karlwilbur] - Group `development` gems to exclude them from CI builds

## 0.9.0 (2020-12-05)

### Additions

* [karlwilbur] - Add tests for:

    - `EasyCaptcha::Captcha`
    - `EasyCaptcha::CaptchaController`
    - `EasyCaptcha::ControllerHelpers`
    - `EasyCaptcha::ModelHelpers`
    - `EasyCaptcha::ViewHelpers`
    - `EasyCaptcha::Espeak` WAV file generation
    - `EasyCaptcha::Generators::InstallGenerator`

* [karlwilbur] - Add `mdl` gem dev dependency; run linters using `bundle exec`
* [karlwilbur] - Add `CHANGELOG`
* [karlwilbur] - Add CodeClimate badges to `README`

### Modifications/Fixes

* [karlwilbur] - Refactor `EasyCaptcha::Espeak` to use `DEFAULT_CONFIG` constant
* [karlwilbur] - Reduce complexity of `EasyCaptcha::Espeak#generate`
* [karlwilbur] - Reduce complexity of `EasyCaptcha#generator`
* [karlwilbur] - Fix logic in `EasyCaptcha::ModelHelper#captcha_valid?`
* [karlwilbur] - Fix bug in audio CAPTCHA generation
* [karlwilbur] - Modify comments and order of config options in initializer template
* [karlwilbur] - Allow `EasyCaptcha#setup` to be called without a block to use defaults
* [karlwilbur] - Lock Simplecov version until CodeClimate Test Reporter is updated

### Deletions/Removals

* [karlwilbur] - Remove deprecation warnings for deprecated methods which were already removed

## 0.8.0 (2020-12-03)

* [karlwilbur] - Rename gem `easy_captcha` -> `kandr-easy_captcha` for publishing to Rubygems
* [karlwilbur] - Convert `README` from rdoc to Markdown
* [karlwilbur] - Fix `EasyCaptcha#respond_to_missing?` logic
* [karlwilbur] - Add pre-commit hook for linting and testing
* [karlwilbur] - Refactor for readibility/maintainability and delinting with help from Rubocop
* [karlwilbur] - Update RSpec tests
* [karlwilbur] - [gemset] Update gems
* [karlwilbur] - [gemspec/Gemfile] Move dev dependencies to `Gemfile`
* [karlwilbur] - [gemfile] Add Rubocop
* [karlwilbur] - [gemspec/gemfile] Update required Ruby and RubyGems versions; set contact info to K&R
* [karlwilbur] - Simplify audio CAPTCHA `generate` method
* [karlwilbur] - Separate CAPTCHA validation and verification
* [karlwilbur] - Add missing `end`
* [karlwilbur] - [gemset] Update gem versions

## 0.7.0 (2020-11-25)

* [karlwilbur] - Update `README.rdoc`
* [karlwilbur] - Remove `.ruby-gemset` file
* [karlwilbur] - Fix `.ruby-version`
* [aditiatama] - Rails 6 compatibility
* [gropher] - Routes fixed
* [mwxjs/unsyn] - bugfixes, logging
* [kopylovvlad] - Rails 5 compatibility
* [brunoporto] - Rails 4.1+ compatibility
* [brunoporto] - fix `font_stroke` and `font_stroke` color and improvements
* [brunoporto] - fix spec
* [forever-sumit] - use `_path` instead of `_url` helper
* [vaucn] - fix misspelling
* [matheusvilela] - fix captcha generation using `rmagick4j` `0.3.8`.

## 0.6

* espeak support for barrier-free support

## 0.5

* (transparent) background support

## 0.4

* generator support

## 0.3

* use generators, optimizations, update licence to same of all my plugins

## 0.2

* cache support for high frequented sites

## 0.1

* init


lib/easy_captcha/generator/default.rb
# frozen_string_literal: true

require 'rmagick'

# EasyCaptcha module
module EasyCaptcha
  # Generator submodule
  module Generator
    # rubocop:disable Metrics/ClassLength
    # default generator
    class Default < Base
      DEFAULT_CONFIG = {
        background_color: '#FFFFFF',
        background_fill: nil,
        background_image: nil,
        blur: true,
        blur_radius: 1,
        blur_sigma: 2,
        font: File.expand_path('../../../resources/captcha.ttf', __dir__),
        font_fill_color: '#333333',
        font_size: 24,
        font_stroke: 0,
        font_stroke_color: '#000000',
        implode: 0.05,
        sketch: true,
        sketch_radius: 3,
        sketch_sigma: 1,
        wave: true,
        wave_amplitude: (3..5),
        wave_length: (60..100)
      }.freeze

      # set default values
      def defaults
        DEFAULT_CONFIG.map do |k, v|
          send("#{k}=", v)
        end
      end

      # Background
      attr_accessor :background_color, :background_fill, :background_image

      # Blur
      attr_accessor :blur, :blur_radius, :blur_sigma

      # Font
      attr_accessor :font_size, :font_fill_color, :font, :font_family, :font_stroke, :font_stroke_color

      # Implode
      attr_accessor :implode

      # Sketch
      attr_accessor :sketch, :sketch_radius, :sketch_sigma

      # Wave
      attr_accessor :wave, :wave_length, :wave_amplitude

      # The CAPTCHA code
      attr_reader :code

      # The CAPTCHA image
      attr_reader :image

      def image_background_color=(val)
        warn  '[DEPRECATION] EasyCaptcha configuration option `image_background_color` is deprecated. ' \
              'Please use `background_color` instead.'
        self.background_color = val
      end

      def image_background_color
        warn  '[DEPRECATION] EasyCaptcha configuration option `image_background_color` is deprecated. ' \
              'Please use `background_color` instead.'
        background_color
      end

      def blur? #:nodoc:
        @blur
      end

      def sketch? #:nodoc:
        @sketch
      end

      def wave? #:nodoc:
        @wave
      end

      # generate image
      def generate(code)
        @code = code
        render_code_in_image
        apply_effects
        create_blob
        set_image_encoding
        free_canvas
        @image
      end

      def apply_effects
        apply_sketch
        apply_implode
        apply_blur
        apply_wave
        apply_crop
      end

      def create_blob
        @image =  if generator_config.background_image.present?
                    create_composite_blob
                  else
                    canvas.to_blob { self.format = 'PNG' }
                  end
      end

      def create_composite_blob
        # Background Random Position
        gravity = [
          Magick::CenterGravity,
          Magick::EastGravity,
          Magick::NorthEastGravity,
          Magick::NorthGravity,
          Magick::NorthWestGravity,
          Magick::SouthGravity,
          Magick::SouthEastGravity,
          Magick::SouthWestGravity,
          Magick::WestGravity
        ].sample

        background = Magick::Image.read(generator_config.background_image).first
        background.composite!(canvas, gravity, Magick::OverCompositeOp)
        background = background.crop(gravity, EasyCaptcha.captcha_image_width, EasyCaptcha.captcha_image_height)
        background.to_blob { self.format = 'PNG' }
      end

      def set_image_encoding
        @image = image.force_encoding 'UTF-8' if image.respond_to? :force_encoding
      end

      def free_canvas
        @canvas.destroy! if canvas.respond_to?('destroy!')
      end

      private

      def apply_blur
        return @canvas unless blur?
        apply_image_blur
        apply_motion_blur
      end

      def apply_image_blur
        return @canvas unless canvas.respond_to?(:blur_image)
        ###
        # https://rmagick.github.io/image1.html#blur_image
        # Parameters:
        #     radius (Float) (defaults to: 0.0) - The radius of the Gaussian operator.
        #     sigma (Float) (defaults to: 1.0) - The standard deviation of the Gaussian operator. Must be non-zero.
        @canvas = canvas.blur_image(
          generator_config.blur_radius,
          generator_config.blur_sigma
        )
      end

      def apply_motion_blur
        return @canvas unless canvas.respond_to?(:motion_blur)
        ###
        # https://rmagick.github.io/image2.html#motion_blur
        # Parameters:
        #     radius (Float) (defaults to: 0.0) - The radius of the Gaussian operator.
        #     sigma (Float) (defaults to: 1.0) - The standard deviation of the Gaussian operator. Must be non-zero.
        #     angle (Float) (defaults to: 0.0) - The angle (in degrees) of the blurring motion.
        @canvas = canvas.motion_blur(
          generator_config.blur_radius,
          generator_config.blur_sigma,
          rand(180)
        )
      end

      def apply_crop
        return @canvas unless canvas.respond_to?(:crop)
        # Crop image because to big after waveing
        ###
        # https://rmagick.github.io/image1.html#crop
        # Parameters:
        #    gravity (Magick::GravityType) - the gravity type
        #    width (Numeric) - width of region
        #    height (Numeric) - height of region
        @canvas = canvas.crop(Magick::CenterGravity, EasyCaptcha.captcha_image_width, EasyCaptcha.captcha_image_height)
      end

      def apply_implode
        ###
        # https://rmagick.github.io/image2.html#implode
        # Parameters:
        #     amount (Float) (defaults to: 0.50) - The precentage to implode the image
        @canvas = canvas.implode(generator_config.implode.to_f) if generator_config.implode.is_a? Float
      end

      def apply_sketch
        return @canvas unless sketch? && canvas.respond_to?(:sketch)
        ###
        # https://rmagick.github.io/image3.html#sketch
        # Parameters:
        #     radius (Float) (defaults to: 0.0) - The radius of the Gaussian operator.
        #     sigma (Float) (defaults to: 1.0) - The standard deviation of the Gaussian operator.
        #     angle (Float) (defaults to: 0.0) - The angle (in degrees) of the sketch lines.
        @canvas = canvas.sketch(generator_config.sketch_radius, generator_config.sketch_sigma, rand(180))
      end

      def apply_wave
        return @canvas unless wave? && canvas.respond_to?(:wave)
        ###
        # https://rmagick.github.io/image3.html#wave
        # Parameters:
        #     amplitude (Float) (defaults to: 25.0) - the amplitude
        #     wavelength (Float) (defaults to: 150.0) - the wave length
        @canvas = canvas.wave(random_wave_amplitude, random_wave_length)
      end

      # rubocop:disable Metrics/AbcSize
      def canvas
        config = generator_config
        @canvas = nil if @canvas.respond_to?('destroyed?') && @canvas.destroyed?
        ###
        # https://rmagick.github.io/image1.html#new
        # Parameters:
        #     cols (Numeric) - the image width
        #     rows (Numeric) - the image height
        #     fill (Magick::GradientFill, Magick::HatchFill, Magick::TextureFill) (defaults to: nil)
        #                    - if object is given as fill argument, background color will be filled using it.
        @canvas ||= Magick::Image.new(
          EasyCaptcha.captcha_image_width,
          EasyCaptcha.captcha_image_height,
          config.background_fill
        ) do |image|
          image.background_color = config.background_color unless config.background_color.nil?
          image.background_color = 'none' if config.background_image.present? || config.background_fill.present?
        end
      end
      # rubocop:enable Metrics/AbcSize

      def generator_config
        @generator_config ||= self
      end

      def random_wave_amplitude
        rand(
          generator_config.wave_amplitude.last - generator_config.wave_amplitude.first
        ) + generator_config.wave_amplitude.first
      end

      def random_wave_length
        rand(
          generator_config.wave_length.last - generator_config.wave_length.first
        ) + generator_config.wave_length.first
      end

      # rubocop:disable Metrics/AbcSize
      def render_code_in_image
        config = generator_config
        # Render the text in the image
        Magick::Draw.new.annotate(canvas, 0, 0, 0, 0, code.gsub(/%/, '\%')) do
          self.gravity     = Magick::CenterGravity
          self.font        = config.font
          self.font_weight = Magick::LighterWeight
          self.fill        = config.font_fill_color
          if config.font_stroke.to_i.positive?
            self.stroke       = config.font_stroke_color
            self.stroke_width = config.font_stroke
          end
          self.pointsize = config.font_size
        end
      end
      # rubocop:enable Metrics/AbcSize
    end
    # rubocop:enable Metrics/ClassLength
  end
end
